src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -95,56 +95,77 @@ EnsureReferenceTablesExistOnAllNodes(void)
+	uint64 shardId = INVALID_SHARD_ID;
+	List *newWorkersList = NIL;
+	const char *referenceTableName = NULL;
+	int colocationId = CreateReferenceTableColocationId();
+
 	/*
-	 * Prevent this function from running concurrently with itself.
-	 *
-	 * It also prevents concurrent DROP TABLE or DROP SCHEMA. We need this
-	 * because through-out this function we assume values in referenceTableIdList
-	 * are still valid.
+	 * Most of the time this function should result in a conclusion where we do not need
+	 * to copy any reference tables. To prevent excessive locking the majority of the time
+	 * we run our precondition checks first with a lower lock. If, after checking with the
+	 * lower lock, that we might need to copy reference tables we check with a more
+	 * aggressive and self conflicting lock. It is important to be self conflicting in the
+	 * second run to make sure that two concurrent calls to this routine will actually not
+	 * run concurrently after the initial check. That is also the reason why we release
+	 * the lock between the two iterations of precondition checks.
 	 *
-	 * We don't need to handle other kinds of reference table DML/DDL here, since
-	 * master_copy_shard_placement gets enough locks for that.
+	 * If after two iterations of precondition checks we still find the need for copying
+	 * reference tables we exit the forloop with the last lock held. This will prevent
+	 * concurrent DROP TABLE and create_reference_table calls so that the list of
+	 * reference tables we operate on are stable.
 	 *
-	 * We also don't need special handling for concurrent create_refernece_table.
-	 * Since that will trigger a call to this function from another backend,
-	 * which will block until our call is finished.
+	 * Since the changes to the reference table placements are made via loopback
+	 * connections we release the final lock held at the end of this function.
 	 */
-	int colocationId = CreateReferenceTableColocationId();
-	LockColocationId(colocationId, ExclusiveLock);
-
-	List *referenceTableIdList = CitusTableTypeIdList(REFERENCE_TABLE);
-	if (referenceTableIdList == NIL)
+	LOCKMODE lockmodes[] = { AccessShareLock, ExclusiveLock };
+	for (int l = 0; l < lengthof(lockmodes); l++)
 	{
-		/* no reference tables exist */
-		UnlockColocationId(colocationId, ExclusiveLock);
-		return;
-	}
+		LockColocationId(colocationId, lockmodes[l]);
 
-	Oid referenceTableId = linitial_oid(referenceTableIdList);
-	const char *referenceTableName = get_rel_name(referenceTableId);
-	List *shardIntervalList = LoadShardIntervalList(referenceTableId);
-	if (list_length(shardIntervalList) == 0)
-	{
-		/* check for corrupt metadata */
-		ereport(ERROR, (errmsg("reference table \"%s\" does not have a shard",
-							   referenceTableName)));
-	}
+		referenceTableIdList = CitusTableTypeIdList(REFERENCE_TABLE);
+		if (referenceTableIdList == NIL)
+		{
+			/* no reference tables exist */
+			for (int ll = l; ll >= 0; ll--)
+			{
```

Comment:
I'm doubtful about `UnlockColocationId` calls. Shouldn't we really need to keep these locks until the end of the transaction block? I mean, if a concurrent transaction does `replicate reference table`, shouldn't it be blocked until this transaction ends even this transaction only read it?

I know, we have that on the main branch as well, but that also looks very interesting to me now.

Reply:
I believe we do, otherwise we serialize on the function call right after a new node has been added.

two functions both enter the checks with the first lock, any of the two will acquire the higher lock and replicate the reference tables. This replication gets committed via a loopback connection to the database. If we were to keep the locks the one that got blocked on the higher lock will need to wait for the first transaction to complete without cause.

On `main` this definitely is a problem, because it didn't get the colocation lock again when adding a table to the colocation group. Causing some metadata inconsistencies that are coincidentally solved with this patch as well.